### PR TITLE
Add report metadata to allow storing/retrieving type-specific values …

### PR DESCRIPTION
**Pull request by wtritch**
…for a Report

The goal is to be able to store the actual aggregate report result count in the Report when it comes back from the aggregate-service AS WELL AS return that value to the front-end so it knows whether or not to enable the "View" link in my "My Reports" view.

### CODE REVIEW

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/Report.java [hunk 1]
```diff
@@ -66,24 +67,17 @@ public AbstractExamReportRequest getReportRequest() {
     }
 
     /**
-     * @return The total number of student chunks for this report
-     */
-    public int getTotalChunkCount() {
-        return totalChunkCount;
-    }
```
Comment (wtritch) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/model/Report.java]: This and the completed chunk count are both type-specific metadata for PDF reports.  I've converted them into metadata.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java [hunk 2]
```diff
@@ -109,40 +121,42 @@ public Report update(final Report report) {
         if (updated != 1) {
             throw new IllegalStateException("Unable to update report: " + report.getId());
         }
+
+        deleteMetadataForReport(report.getId());
+        for (final Map.Entry<String, String> entry : report.getMetadata().entrySet()) {
+            insertMetadataForReport(report.getId(), entry.getKey(), entry.getValue());
+        }
```
Comment (wtritch) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java]: When updating a Report instance I wipe and re-insert it's metadata.  This should not happen often.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java [hunk 3]
```diff
@@ -109,40 +121,42 @@ public Report update(final Report report) {
...
 
-        return template.query(query,
+        return saturateMetadata(template.query(query,
                 new MapSqlParameterSource()
                         .addValue("user", user)
                         .addValue("report_ids", reportIds),
-                (row, rowNum) -> asReport(row)
-        );
+                (row, num) -> asReport(row)
+        ));
```
Comment (wtritch) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java]: All the find* methods now saturate the found reports with their metadata.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java [hunk 4]
```diff
@@ -199,17 +237,45 @@ private String uriAsString(final URI value) {
...
+                    if (reportMetadata == null) {
+                        return report;
+                    }
+
+                    return report.copy()
+                            .metadata(reportMetadata)
+                            .build();
+                })
+                .collect(Collectors.toList());
+    }
```
Comment (wtritch) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java]: Given a collection of reports I find the metadata for all reports in the collection then map it onto the reports using copy/set/build.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java (hunk 2 above)
Comment (malaffoon) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java]: You do this just so i have something to add, don't you?
s/it's/its/g

File: report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepositoryIT.java [hunk 5]
```diff
@@ -170,4 +182,45 @@ public void itShouldFindReportFrom30DaysAgo() {
...
+                Integer.class);
+        assertThat(metadataCount).isEqualTo(1);
+
+        repository.deleteIds(Collections.singleton(original.getId()));
+
+        metadataCount = template.queryForObject("SELECT count(*) from user_report_metadata",
+                new MapSqlParameterSource(),
+                Integer.class);
+        assertThat(metadataCount).isEqualTo(0);
+    }
```
Comment (wtritch) [inline report-processor/src/test/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepositoryIT.java]: Proving that updating and deleting still appear to act on a single Report instance to the world outside the repository.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java (hunk 2 above)
Comment (wtritch) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java]: I have to make sure you're paying attention.  It's a habit.

File: report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java [hunk 6]
```diff
@@ -199,17 +237,45 @@ private String uriAsString(final URI value) {
...
     }
+
+    private RowCallbackHandler asReportMetadata(final Map<Long, Map<String, String>> metadata) {
+        return (row) -> {
+            final Long reportId = row.getLong("report_id");
+            final Map<String, String> reportMetadata = metadata.computeIfAbsent(reportId, (id) -> new HashMap<>());
+            reportMetadata.put(
+                    row.getString("name"),
+                    row.getString("value")
+            );
```
Comment (esotrope) [inline report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java]: looks like a one-liner to me : j

Comment (wtritch) [reply report-processor/src/main/java/org/opentestsystem/rdw/reporting/processor/repository/JdbcReportRepository.java]: I tried that, but it felt all unwieldy n' stuff.

**Pull request merged**